ratik/src/intTest/java/org/opencadc/ratik/InventoryValidatorTest.java

```diff
@@ -0,0 +1,759 @@
+        Assert.assertEquals("metaChecksum mismatch", artifact.getMetaChecksum(), localArtifact.getMetaChecksum());
+    }
+
+    /** discrepancy: artifact not in L && artifact in R
+     *
+     * explanation4: L==global, new Artifact in R, pending/missed changed Artifact event in L
+     * evidence: Artifact in local db but siteLocations does not include remote siteID
+     * action: add siteID to Artifact.siteLocations
+     *
+     * L == global
+     * before: Artifact in L & R, R siteID not in L Artifact.siteLocations
+     * after:  Artifact in L and R siteID in Artifact.siteLocations
+     */
+    @Test
+    public void explanation4_ArtifactNotInL_LocalIsGlobal() throws Exception {
+        UUID remoteSiteID = this.remoteEnvironment.storageSiteDAO.list().iterator().next().getID();
+
+        Artifact artifact = new Artifact(URI.create("cadc:INTTEST/one.ext"), TestUtil.getRandomMD5(),
+                                         new Date(), 1024L);
+        this.remoteEnvironment.artifactDAO.put(artifact);
+        this.localEnvironment.artifactDAO.put(artifact);
+
+        try {
+            System.setProperty("user.home", TMP_DIR);
+            InventoryValidator testSubject = new InventoryValidator(this.localEnvironment.daoConfig, TestUtil.LUSKAN_URI,
+                                                                    new IncludeArtifacts(),null,
+                                                                    true);
+            testSubject.run();
+        } finally {
+            System.setProperty("user.home", USER_HOME);
+        }
+
+        Artifact localArtifact = this.localEnvironment.artifactDAO.get(artifact.getID());
+        Assert.assertNotNull("local artifact not found", localArtifact);
+        Assert.assertEquals("metaChecksum mismatch", artifact.getMetaChecksum(), localArtifact.getMetaChecksum());
+        Assert.assertTrue("artifact does not contains remote site location",
+                          localArtifact.siteLocations.contains(new SiteLocation(remoteSiteID)));
+    }
+
+    /**
+     * discrepancy: artifact.uri in both && artifact.id mismatch
+     *
+     * explanation1: same ID collision due to race condition that metadata-sync has to handle
+     * evidence: no more evidence needed
+     * action: pick winner, create DeletedArtifactEvent for loser, delete loser if it is in L
+     *
+     * winner always in remote?
+     * before: Artifact.uri in L && R, L Artifact.id != R Artifact.id
+     * after:  Artifact not in L, DeletedArtifactEvent in L
+     */
+    @Test
+    public void artifactUriCollision_LocalIsStorage() throws Exception {
+        artifactUriCollision(false);
+    }
+
+    @Test
+    public void artifactUriCollision_LocalIsGlobal() throws Exception {
+        artifactUriCollision(true);
+    }
+
+    public void artifactUriCollision(boolean trackSiteLocations) throws Exception {
+        URI contentCheckSum = TestUtil.getRandomMD5();
+        UUID localID = UUID.randomUUID();
```

Comment:
In this discrepancy, the "pick winner" is not as simple as "remote wins", which is what you have here. You have to pick winner the same way that fenwick does because either fenwick or ratik could encounter this and they have to resolve it the same way.

So there are nominally two scenarios hiding here - one where the local artifact wins and one where the remote artifact wins - and the results are different. The checks need to verify that the expected winner ends up in L and the DAE is created for the loser. You don't have to explicitly check that the loser got deleted because DB constraints won't let the winner be inserted in that case

Reply:
Spend a long time going through not many lines in fenwick that determine which Artifacts wins. Updated the tests to pick the same winner.